halo2-base/src/gates/flex_gate.rs

```diff
@@ -787,6 +789,35 @@ pub trait GateInstructions<F: ScalarField> {
         ctx.get(-2)
     }
 
+    /// Constrains that a cell value is even and returns `1` if `a is even`, otherwise `0`.
+    /// * `ctx`: [Context] to add the constraints to
+    /// * `a`: [QuantumCell] value to be constrained
+    fn is_even(&self, ctx: &mut Context<F>, a: AssignedValue<F>) -> AssignedValue<F> {
+        let x = fe_to_biguint(a.value());
+        let (div, rem) = x.div_mod_floor(&BigUint::from(2u64));
+
+        let one_minus_rem = biguint_to_fe(&(BigUint::from(1u64) - &rem));
+        let div = biguint_to_fe(&div);
```

Comment:
`div` needs to be range checked: you can use `RangeInstructions::get_last_bit` function as a stand-in for `is_even`